### PR TITLE
fix: Improve Sentry logging for tests

### LINKED ISSUE #1496: Sentry should not report errors during tests + general sentry enhancements
Some of flank unit tests throw errors (intentionally) to verify exception handling logic https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/util/UtilsTest.kt

There is a possibility there are more places similar which also throws errors.
Currently, Sentry is disabled only when `@RunWith(FlankTestRunner::class)` is used (it sets `useMock` flag to true which prevents initializing Sentry)

Some 'test' errors reported in Sentry
https://sentry.io/organizations/flank/issues/?project=5514444&query=is%3Aunresolved+%2ATest%2A&statsPeriod=14d
Screenshot 2021-01-14 at 08 40 06

There are also some improvements that could be done within this issue

-------------

#### TODO list
1. Find a way to disable Sentry for tests, ideas:
    * adding `@RunWith(FlankTestRunner::class)` everywhere is rather not an option, maybe another runner?
    * reuse logic (with small refactor) for flank properties https://github.com/Flank/flank/blob/7a2a9dc2adc3f123fc72eb03e7e9c1121a93895c/build.gradle.kts#L31-L40 https://github.com/Flank/flank/blob/7a2a9dc2adc3f123fc72eb03e7e9c1121a93895c/common/src/main/kotlin/flank/common/config/FlankProperties.kt#L37 and do not initialize Sentry if `isTest() == true`
    * ???
2. Test `should notify bugsnag if non-related flak error occurred` should be fixed. This commit https://github.com/Flank/flank/commit/9bf7b237d2787c82df8883f86106e67e03c1b8e4#diff-ee8e699d4239457ff09d8ddeff4b4dc087c35685b68226dca3d8d0c8614f02feL266-L268 made test to check actually nothing (there should be verification if sentry is notified when no flank related error is thrown). Seems like that test was 
(partially) broken before, in https://github.com/Flank/flank/commit/c9e349a7bc2aad0f81b625c2af53eda81adaafc3
3. search and fix/replace all remaining `bugsnag` words (like in the test mentioned above)
4. check if other 'exception handling tests` work correctly
5. remove white space (or is it intentionally?) https://github.com/Flank/flank/blob/7a2a9dc2adc3f123fc72eb03e7e9c1121a93895c/test_runner/src/main/kotlin/ftl/util/CrashReporter.kt#L15 that causes strange `flank.version` tag 
Screenshot 2021-01-14 at 09 05 20
6. clear Sentry log from non-production errors

#### DoD
Once work is done no new 'test errors' are reported to Sentry

### PR DESCRIPTION
**Pull request by piotradamczyk5**
Fixes #1496 

Tests crashes are not reported to Sentry during performing tests

## Checklist

- [x] Disable logging for test
- [x] Fix test `should notify bugsnag if non-related flak error occurred`
- [x] Replace/remove bugsnag reference
- [x] Check and refactor exception handling tests
- [x] Remove white space from `" flank.version"`
- [x] clear Sentry log from non-production errors

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt [hunk 1]
```diff
@@ -3,18 +3,23 @@ package ftl.run.exception
...
-import ftl.util.captureError
+import ftl.util.report
 import kotlinx.coroutines.runBlocking
 import kotlin.system.exitProcess
 
 fun withGlobalExceptionHandling(block: () -> Int) {
+    withGlobalExceptionHandling(block, { exitProcess(it) })
+}
+
+// Overloading this function make tests easier to perform and tests exit code properly
```
Comment (Sloox) [inline test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt:14]: Usually we dont have comments like this.
Regardless the wording is strange.
```suggestion
// Overloading this function makes tests easier to implement and exit with the correct exit code
```

File: test_runner/src/test/kotlin/ftl/util/UtilsTest.kt [hunk 2]
```diff
@@ -180,64 +176,76 @@ class UtilsTest {
...
-        val block = { throw FlankTimeoutError(mapOf("anyMatrix" to mockk(relaxed = true)), "anyProject") }
-        // when
-        withGlobalExceptionHandling(block)
-        // then
-        coVerify(exactly = 1) { cancelMatrices(any(), any()) }
+        val exception = FlankTimeoutError(mapOf("anyMatrix" to mockk(relaxed = true)), "anyProject")
+
+        // when-then
+        withGlobalExceptionHandling(
+            { throw exception },
```
Comment (Sloox) [inline test_runner/src/test/kotlin/ftl/util/UtilsTest.kt:213]: hmmm shall we enable ktlint for tests. I am not liking this styling. ? WDYT?

**Review by Sloox — CHANGES_REQUESTED** [review]

**Review by pawelpasterz — DISMISSED** [review]
When I run `./gradlew clean build` from PR's branch I can see 2 error are reported to sentry
Screenshot 2021-01-18 at 08 12 11
[**ftl.MainTest**](https://sentry.io/organizations/flank/issues/2160663967/?end=2021-01-18T23%3A59%3A59&project=5514444&query=is%3Aunresolved&start=2021-01-18T00%3A00%3A00&utc=true)
[**ftl.util.UtilsTest**](https://sentry.io/organizations/flank/issues/2160661411/?end=2021-01-18T23%3A59%3A59&project=5514444&query=is%3Aunresolved&start=2021-01-18T00%3A00%3A00&utc=true)

File: test_runner/src/test/kotlin/ftl/util/UtilsTest.kt (hunk 2 above)
Comment (piotradamczyk5) [reply test_runner/src/test/kotlin/ftl/util/UtilsTest.kt:213]: I will do it in separate PR

File: build.gradle.kts [hunk 3]
```diff
@@ -33,6 +33,7 @@ subprojects {
     afterEvaluate {
         if (tasks.findByName("test") != null) {
             tasks.test {
+                systemProperty("runningTests", true)
                 systemProperty("useDefaultProperties", "")
```
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply build.gradle.kts:37]: removed

**Comment by piotradamczyk5** [comment]
> When I run `./gradlew clean build` from PR's branch I can see 2 error are reported to sentry
> Screenshot 2021-01-18 at 08 12 11
> [**ftl.MainTest**](https://sentry.io/organizations/flank/issues/2160663967/?end=2021-01-18T23%3A59%3A59&project=5514444&query=is%3Aunresolved&start=2021-01-18T00%3A00%3A00&utc=true)
> [**ftl.util.UtilsTest**](https://sentry.io/organizations/flank/issues/2160661411/?end=2021-01-18T23%3A59%3A59&project=5514444&query=is%3Aunresolved&start=2021-01-18T00%3A00%3A00&utc=true)

nice catch, fixed!

**Review by Sloox — APPROVED** [review]

File: test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt [hunk 4]
```diff
@@ -3,18 +3,27 @@ package ftl.run.exception
...
 
 fun withGlobalExceptionHandling(block: () -> Int) {
+    withGlobalExceptionHandling(block) {
+        closeCrashReporter()
+        exitProcess(it)
+    }
+}
+
+// Overloading this function makes tests easier to implement and exit with the correct exit code
+fun withGlobalExceptionHandling(block: () -> Int, exitProcessFunction: (Int) -> Unit) {
```
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt:19]:
```suggestion
internal fun withGlobalExceptionHandling(block: () -> Int, exitProcessFunction: (Int) -> Unit) {
```

File: test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt [hunk 5]
```diff
@@ -82,3 +91,8 @@ private fun printError(string: String?) = System.err.println(string)
 private fun SavedMatrix.logError() {
     logLn("Matrix is $state")
 }
+
+fun closeCrashReporter() {
```
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt:95]: it should be in different file. I will move it

File: test_runner/src/test/kotlin/ftl/util/UtilsTest.kt (hunk 2 above)
Comment (jan-gogo) [reply test_runner/src/test/kotlin/ftl/util/UtilsTest.kt]: @Sloox what is wrong with this styling? IMO it's better than
```
{
    throw exception
},
```
cause it's shorter

File: test_runner/src/test/kotlin/ftl/util/UtilsTest.kt [hunk 6]
```diff
@@ -247,66 +255,61 @@ class UtilsTest {
 
         val block = { throw FlankGeneralError(message) }
 
-        // will
-        exit.expectSystemExitWithStatus(GENERAL_FAILURE)
-        exit.checkAssertionAfterwards {
+        // when
+        withGlobalExceptionHandling(block) {
```
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply test_runner/src/test/kotlin/ftl/util/UtilsTest.kt:259]: fixed

File: test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/run/exception/ExceptionHandler.kt:95]: fixed

**Comment by adamfilipow92** [comment]
WDYT about disable sentry reporting on debug?

File: test_runner/src/test/kotlin/ftl/util/UtilsTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (jan-gogo) [reply test_runner/src/test/kotlin/ftl/util/UtilsTest.kt]: I agree it's not necessary to store a lambda in value (block) if it is used only once.

**Comment by piotradamczyk5** [comment]
> WDYT about disable sentry reporting on debug?

done, nice idea!

**Review by adamfilipow92 — APPROVED** [review]
- [X] Unit tests not report errors
- [X] Code looks good
- [X] When debugging reporting errors is turned off

**Pull request merged**